### PR TITLE
Step 2 RTC, removes ModelDB and deduplicates the content

### PR DESCRIPTION
**Pull request by hbcarlos**
This PR introduces a new approach for RTC, in line with the old ModelDB, trying to minimize as much as possible backward-incompatible changes. I'm replacing the shared models introduced in #10118 with completely redesigned new ones based on ModelDB API. Instead of having an abstraction for Yjs that at the same time acts as a schema for the document, this new approach provides a more flexible abstraction for Yjs, leaving the DocumentModel as a schema for each document as it was before introducing RTC in #10118.

#### Pros:
* It is much simple and removes code duplication since we don't need to have two models for each document. For example, at the moment, to represent a notebook, we need to create/maintain the NotebookModel and the YNotebook. For a plain-text file, we need the DocumentModel and the YDocument. These Y models have almost the same API as the NotebookModel or DocumentModel, duplicating most logic.

* We keep the concept of observables. For example, two essential properties in the NotebookModel are the CellList and the metadata. Every extension that interacts with the CellList or the metadata requires these objects from the notebook model and forgets about the notebook by listening and applying changes directly to the observable. This approach maintains the concept of observables, avoiding breaking this pattern.

* It is a better abstraction for Yjs. We were still using some Yjs specifics around different packages with the actual approach.

#### Cons:
* Now that we are loading/saving documents from the back-end, we need to sync the front-end and back-end with the schema used for a document. With this approach, in the case of complex documents like the notebook, it is more difficult because we don't create the structure of the notebook in a single file. For example, the list of cells called `cells` is created in the CellList class instead of the NotebookModel class.

## References
This PR depends on:
* https://github.com/jupyterlab/jupyterlab/pull/11640

This PR replaces:
* https://github.com/jupyterlab/jupyterlab/pull/11601
* https://github.com/jupyterlab/jupyterlab/pull/11790

## Code changes
The most significant changes are:
* ModelDB is no longer present in the `DocumentModel` but in the `IContext` and `IModelFactory` if someone wants to create non-collaborative documents.
* Every property in a document model that was an IObservable now is an IShared with a similar API. The most significant differences between IObservable and IShared are the signals.
    Changing the signals is necessary because the observables trigger one signal per change, and with RTC, we need to send multiple changes per signal. I decided to adopt the Delta format used in Yjs because it is easier to translate the signals emitted by Yjs, but we can use another format.
    * IObservableString -> ISharedString
    * IObservableList -> ISharedList
    * IObservableMap -> ISharedMap

## Backwards-incompatible changes
<details>

### attachments
* IAttachmentsModel.IOptions: 
    * -modelDB?: `IModelDB`
    
### cells:
* ICellModel
    * +-metadata `IObservableJSON` -> `ISharedMap<JSONObject>`
    * +-sharedModel `IObservableJSON` -> `ISharedMap<ISharedType>`
* CellModel
    * -switchSharedModel
    * +_onValueChanged
    * IOptions:
        * -modelDB?: `IModelDB`
* ICodeCellModel
    * -sharedModel: `ISharedCodeCell`
* CodeCellModel
    * -switchSharedModel
    * -sharedModel: `ISharedCodeCell`

### codeeditor:
* CodeEditor.IModel
    * +-value: `IObservableString` -> `ISharedString`
    * -sharedModelSwitched
    * -switchSharedModel
    * -sharedModel
    * -modelDB: `IModelDB`
* CodeEditor.Model
    * +-`constructor(options?: Model.IOptions)` -> `constructor(options: Model.IOptions)`
    * +-value: `IObservableString` -> `ISharedString`
    * -sharedModelSwitched
    * -switchSharedModel
    * -sharedModel
    * -modelDB: `IModelDB`
* CodeEditor.Model.IOptions:
    * -id?: `string`
    * -value?: `string`
    * -modelDB?: `IModelDB`
* JSONEditor
    * +-source `IObservableJSON | null` -> `ISharedMap<JSONObject> | null`

### docprovider:
* IDocumentProviderFactory.IOptions
    * -ymodel

### docregistry:
* DocumentModel
    * +-constructor `constructor(languagePreference?: string, modelDB?: IModelDB)` -> `constructor(languagePreference?: string, sharedDoc?: ISharedDoc)`
* DocumentRegistry.IModel
    * -sharedModel
    * -modelDB: `IModelDB`
* DocumentRegistry.ICodeModel
    * -sharedModel

### notebook:
* CellList: `IObservableUndoableList<ICellModel>` -> `ICellList`
    
* INotebookModel
    * +-cells: `IObservableUndoableList<ICellModel>` -> `ICellList`
    * +-metadata: `IObservableJSON` -> `ISharedMap<JSONObject>`
    * -sharedModel
* NotebookModel
    * +-cells: `IObservableUndoableList<ICellModel>` -> `ICellList`
    * +-metadata: `IObservableJSON` -> `ISharedMap<JSONObject>`
    * -sharedModel
    * -modelDB: `IModelDB`
* NotebookModel.IOptions:
    * -modelDB?: `IModelDB`
* NotebookModel.IContentFactory:
    * +-clone: `clone(modelDB: IModelDB): IContentFactory;` -> `clone(sharedDoc: ISharedDoc): IContentFactory;`
    * -modelDB: `IModelDB | undefined`
* NotebookModel.ContentFactory:
    * +-clone: `clone(modelDB: IModelDB): IContentFactory;` -> `clone(sharedDoc: ISharedDoc): IContentFactory;`
    * -modelDB: `IModelDB | undefined`
* NotebookModel.IContentFactory.IOptions:
    * -modelDB: `IModelDB`
</details>

## TODO:

- [x] Uncomment move in CellList after #11640
- [x] Improve the initialization after #12360
- [ ] Define an API for Awareness
- [x] Fix tests
- [x] Create collaborative tests
- [ ] Open a PR to deprecate backward-incompatible changes in v3.4
    
## Questions:
* Should we share attachments in the notebook?
* What format should we use for the signals?

### CODE REVIEW

**Comment by dmonad** [comment]
This work came as kind of a surprise to me. I feel this PR is a massive step back from what we accomplished last year.

I'm trying to give everyone an understanding of what is happening in this PR. 

## This PR will remove the shared-model approach that we popularized last year as the new API endpoint for Jupyter Notebooks.

It was my understanding that we wanted to introduce a consistent, easy-to-use API with the shared-model implementation.

We moved away from the modeldb approach because it is notoriously difficult to understand how the shared document it is used by all the different packages (notebook model, cell model, cells, editor model, docregistry, ...). All of these packages are responsible for rendering the from the mostly untyped modeldb store to the DOM and handling events.  The main takeaway for me from my first PR (which also replaced modeldb with Yjs) was that these packages should not be responsible for maintaining the data that is synchronized between the server and many peers. Something like a schema was needed (which don't exist yet for CRDTs as they are a major subject of research). @echarles and I came up with the approach to put the shared data into a single package where we can control how data is represented in Yjs, how to handle conflicts, and how to ensure that the exposed data is used. This allows us to separate concerns (model representation & how to render the model) and would also make it easy to switch from Yjs to a different CRDT in the future without changing the whole codebase.

I wrote a whole blog post about how great this approach is and I've been recommending this to other companies that want to write complex applications with Yjs.

The main advantages of the original shared-model approach are..
* Having a single place where we model shared data.
* A single easy-to-use API for plugins to manipulate the notebook and listen to changes.
* Ensuring that the shared model conforms to some sort of schema. Yjs is untyped and everyone is potentially able to write anything to Yjs.
* Potentially being able to replace Yjs with a different CRDT in the future.
* Ensuring that plugins don't insert invalid data by having a typed API.
* Being able to throw an error message when two clients use different versions of "shared-model" that are incompatible.
* Being able to ensure backward compatibility while we evolve the schema.

I'm particularly excited about the fact that others can use `shared-model` to implement their custom interface on-top of it. All clients that use the same `shared-model` are able to collaborate with each other.

Since we maintain all the Yjs-internal logic in a single package, it makes it fairly easy to write a python implementation of the shared-model. In fact, a python implementation of a shared-model is needed because we want to avoid critical issues in collaborative model (see https://github.com/jupyterlab/jupyterlab/issues/12154#issuecomment-1088358925). We need to be able to have a common understanding of how we insert data into the Yjs document. The shared model serves as our schema, our API, and our common understanding so we avoid incompatibilities between versions.

## This PR will basically re-introduce modeldb under a different name (which unfortunately is also called "shared models"..).

This is basically my very first approach from [February last year](https://github.com/jupyterlab/jupyterlab/pull/9785), but it adds an unnecessary abstraction around Yjs. 

This approach is fundamentally untyped and will make all the different packages again responsible rendering the dom and __also__ maintaining the structure of the shared Yjs document. These concerns should be seperated. I highly recommend that JupyterLab maintains shared models in single place which handles all the complicated edge-cases of concurrency. Third-party packages should not get in contact with the CRDT itself because they could insert invalid data (or data that no other client can understand).

Another point: This PR, again, puts the responsibility of initializing the document content on the `constructors` of the different editor components (See e.g. [here](https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-97b5f1c7b9bbe5df6760028afd7dfdcfec36be10f54eb1e1445c3331e2232c7bR225)).  Instead, the initialization part should be done when creating the model (e.g. the cell component). I.e. there are two ways to create a cell component: 1. We create a cell component and fill it with content. 2. We receive a cell component from a remote peer with initialized content. We don't want to run both edge-cases through the same constructor. The original shared-model approach solves this beautifully by having a separate factory method to create cells with initialized content. I talked on many different occasions that using a single constructor for everything is unclean and can lead to deeper issues.

The API to manipulate notebooks is now again spread over ~20 different packages. Alternatively you can manipulate the Yjs document directly, but you'd have no typing information and you need to figure out yourself how the Jupyter notebook is represented in Yjs.

The "view" component should have no saying in how the shared data is modeled. This can work in smaller projects, but in larger projects, this lack of abstraction is really hard to maintain. Places that initialize the model (just an extract):

* attachments - constructor https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-db040a533cc47dd3041d1d125065fc7fc0f8279a8259879a446b01472719b1f3R140
* cells - Case in point, you forgot to initialize "editable" in constructor but already used it in different places e.g. https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-352916c4cc928b1298618fffc0424a71ee8816925f3a36c77d8593057e7b6f89R311
* codeeditor - constructor defines value https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-97b5f1c7b9bbe5df6760028afd7dfdcfec36be10f54eb1e1445c3331e2232c7bR226
* codemirror - inherits from codemirror and defines `mimetype`  https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-9705cdaacb4660fc2de7eaabf424d6c24b672acc499292fec723a779be32fbc7R135
* notebook/panel - How do we know that `language_info` is defined and what would happen if we rename the untyped property 'language_info'? This was just so much nicer in the original typed approach.. https://github.com/jupyterlab/jupyterlab/pull/9785/files#diff-8d0179dcf34ad7f95ff4b4e2d78b4da8faa8d24d5a7069f63a78402c73de4629R223
* This really was just an extract of potential problems..

Now, if I would attempt to extract the content from the Yjs document (e.g., on the python server) it would be impossible for me to understand which properties should be defined and how they are named. I would have to dig through the whole Jupyter codebase. Hence we should put everything into a single module to make this process easier.

## This PR contains no advantages

I don't see any apparent advantages of using Yjs like this over the original "shared-model" approach.

> It is much simple and removes code duplication since we don't need to have two models for each document. For example, at the moment, to represent a notebook, we need to create/maintain the NotebookModel and the YNotebook. For a plain-text file, we need the DocumentModel and the YDocument. These Y models have almost the same API as the NotebookModel or DocumentModel, duplicating most logic.

Yes, you removed modeldb. That is something we need to do. However, you actually added much more code in this PR than you removed + you increased complexity (IMHO). 
image


> We keep the concept of observables. For example, two essential properties in the NotebookModel are the CellList and the metadata. Every extension that interacts with the CellList or the metadata requires these objects from the notebook model and forgets about the notebook by listening and applying changes directly to the observable. This approach maintains the concept of observables, avoiding breaking this pattern.

We never removed observable, and the original shared model approach in fact exposes a very similar API that is easier to use. We don't have to remove the current observables until everyone migrated to using shared-model. The new observables introduced by the original shared-models are actually favorable and expose more relevant information using a typed interface. We can keep the old observables for a time, but we can remove them eventually once everyone migrated to shared-models.

> It is a better abstraction for Yjs. We were still using some Yjs specifics around different packages with the actual approach.

I don't see what you mean. We only access Yjs-internals when binding to an editor and syncing to the backend. It was always possible to implement the original shared-model with a different CRDT and then only change the few places where we access Yjs directly. The idea was to limit the usage of Yjs specific features which we successfully did.

----

The approach implemented in this PR can work, however, I have a lot of concerns and feel that it is a massive step back. However, I will not block this PR because I have less of a stake in JupyterLab. Over the coming year, I will probably fade out of the project (once funding for me stops) and will only be available if you have questions. I won't be the person who maintains the RTC part of Jupyterlab (however, again, always available for questions). So whether you want to go with these changes is really up to the people that maintain them.

**Comment by echarles** [comment]
Thank you @hbcarlos and @dmonad for working towards a better collaborative jupyterlab.

It has been great brainstorming and working together last year on the first iteration and making real the idea of `shared-model` package that could be used by all other jupyterlab packages, but also by 3rd party applications. We have laid down the foundation for that, and we knew it was not perfect and remaining work was left to remov the dual model approach (the modeldb and the shared-model). The gist of the design was to have nice API hiding the complexity of CRDT and also allowing the hide the complex operations and subtle edge cases that collaborative CRDT present. Having such an API would also allow to be immune to the evolution of the Y.js APIs and allow supporting upcoming libraries, wherever it is Y.js or others.

@hbcarlos I'd love to get more insights on the attention points raised by @dmonad in terms of `initialisation in the constructor`, `typings` and `responsibility for manipulating the Y.js document directly`. Also, are you really removing all usage of modelDB? The Context still create a model passing a modelDB attribute if a modelDB is given via the option..

```
      this._model = this._factory.createNew(
        lang,
        this._modelDB,
        this._sharedDoc,
        false
      );
```

I still need to make sense of the additional attribute on the model `isDocument: boolean;` as the need for the `underlyingModel`.

I see a TODO `Create collaborative tests`. One more general comment from me would be to start creating asap that monkey test suite (random usage collaboration) to challenge the current implementation vs any new one to ensure we don't hit edge cases. The current test suites are mono-user and do no allow to anticipate any real-life issues.

**Comment by hbcarlos** [comment]
Hi @echarles. Thank you for your feedback, and sorry for the late response. I wrote down a more detailed explanation for your comment:

> Also, are you really removing all usage of modelDB? The Context still create a model passing a modelDB attribute if a modelDB is given via the option..

Yes, I am. I have removed ModelDB from the default documents (DocumentModel, and NotebookModel). I left ModelDB in context and the factory as optional in case extension developers want to keep using it for their custom non-collaborative documents, or they prefer to wait before updating to the shared model.

> responsibility for manipulating the Y.js document directly

It is the same as Kevin's approach, the difference is that he created a new model for each document (YDocument, YNotebook) and I'm using the existing one (DocumentModel, NotebookModel). My approach is much simpler since we don't need to maintain two models for each document.

> I still need to make sense of the additional attribute on the model `isDocument: boolean;` as the need for the `underlyingModel`.

The `isDocument: boolean;` is not strictly necessary, but it is better if we explicitly indicate what schema are we expecting from the IModel. This is the same as Kevin is doing with `isStandalone` in the shared models. The reason is that the `codeeditor.IModel` is the underlying model that represents the source (code) of each document, the problem is that the notebook has multiple sources (each cell is a different IModel) so, the schema of a plain text document is different from the schema of a Notebook and we store the attribute `source` in a different place on the Y.Doc.

Regarding `underlyingModel`, it is just because we need to access the YType to create bindings. For example, the Y-CodeMirrorBinding needs a YText.
In other places, for example, the CellList, when creating a mapping (we can not store an object in Yjs, we need a mapping) we need to use the YType as a key for storing the CellModel because Yjs uses references to the JSON objects or YTypes stored on a YType so these objects will always be the same and we can use them as key for the mapping. Instead, we can not use an attribute of the content (for example the cell id) because once we remove an element from a YType, its content is no longer available so we can not get the `id` to then remove the mapping (the CellModel).

> I see a TODO Create collaborative tests. One more general comment from me would be to start creating asap that monkey test suite (random usage collaboration) to challenge the current implementation vs any new one to ensure we don't hit edge cases. The current test suites are mono-user and do no allow to anticipate any real-life issues.

That's my idea, the problem is that I don't know how to make it generic for any implementation. I was looking into the Yjs protocol to create yjs-specific tests that for example, drop messages, disconnect clients, open two clients at the same time...

> initialisation in the constructor

Kevin is right on that point and I already changed that.

> typings

There is typing information in a more generic way as I believe it should be since every extension adds attributes that we don't know about. For example, the metadata is a `Map<JSONObject>`. This prevents developers for adding objects that Yjs can not understand while allowing them to add any attribute they need.

**Comment by echarles** [comment]
@hbcarlos Thx for the answers and latest commits. I will be happy to review and test, maybe once @dmonad will have given, if he has cycles for that, more feedback on the latest changes.

**Comment by dmonad** [comment]
> It is the same as Kevin's approach, the difference is that he created a new model for each document (YDocument, YNotebook) and I'm using the existing one (DocumentModel, NotebookModel). My approach is much simpler since we don't need to maintain two models for each document.

My feedback is still that this is a step back. The database schema (in this case how we represent JupyterNotebooks in a Yjs document) should be separated from the application state (how we model application state: DocumentModel, NotebookModel, CodeCellModel, CodeEditorModel, CodemirrorModel, ...). It does make sense to separate these two and think about them differently.

Writing Jupyters' shared-model was a huge revelation for me. Separating the concern of application state from database schema makes maintenance and further development much easier. Furthermore, it finally gives extension developers a clean API to work against.

I still don't understand what this PR actually achieves. I think there is a misguided sense that Yjs should be integrated into the application model because that's the "idiomatic approach". I don't see that.  Furthermore, this PR clearly increases complexity. Now it has 2200 additional lines of code.

Maybe somebody can explain to me why this approach is easier.

**Comment by echarles** [comment]
> Writing Jupyters' shared-model was a huge revelation for me. Separating the concern of application state from database schema makes maintenance and further development much easier. Furthermore, it finally gives extension developers a clean API to work against.

The `shared-model` idea has matured from various implementation iterations and from brainstorming during in-person calls. We also said that we would need to take next steps on it.

> Maybe somebody can explain to me why this approach is easier.

I have the feeling that in-person calls are needed to truly understand what the next steps should be and how this PR addresses them. These discussions should occur with people implementing the code to have the best chance to give concrete results. I haven't touched RTC in jupyterlab since some time. Based on the work done last year, I may be able to bring help.

@hbcarlos would you take a lead to schedule a meeting with @dmonad and myself? Anyone else welcome to join.

**Comment by SylvainCorlay** [comment]
> Furthermore, this PR clearly increases complexity. Now it has 2200 additional lines of code.

The increase in the number of lines of code is due to the addition of tests, documentation, and doc strings.

> My feedback is still that this is a step back. The database schema (in this case how we represent JupyterNotebooks in a Yjs document) should be separated from the application state (how we model application state: DocumentModel, NotebookModel, CodeCellModel, CodeEditorModel, CodemirrorModel, ...). It does make sense to separate these two and think about them differently.
> Writing Jupyters' shared-model was a huge revelation for me. Separating the concern of application state from database schema makes maintenance and further development much easier. 

After talking about this with several people in the team, I am clearly in favor of Carlos' approach in this PR.

Having a separate package where data models are held will force us to modify two packages whenever changing one of these models. I don't see what the value add is.

Also, putting these shared models of unrelated document types in a single package is definitely bad as you can't change the schema of one of them without bumping the versions for all of them.

> Maybe somebody can explain to me why this approach is easier.

We need to remove the duplication of state between e.g. the notebook model and the shared notebook model. There are two ways to do this: the one proposed in this PR, or removing the notebook model altogether (replacing DocumentModel and NotebookModel with YDocument and YNotebook). The second approach is a *lot* more disruptive IMO.

Beyond that, there are some other benefits to Carlos' approach

 - In your approach, the notebook and cell metadata cannot be accessed directly, but one can retrieve a copy typed as `nbformat.INotebookMetadata` with a getter, or set it globally with a setter. This prevents using tools such as the generic JSON editor of JupyterLab to edit the notebook or cell metadata in place. This also forces one to invoke the getter on any single-attribute change to read the new value.
 - Many extensions add attributes to the metadata that are not part of the notebook format, and several listen to changes of these values. The way shared models work with metadata prevents that (or you would need to listen to the entire metadata change and invoke the getter each time).

Regarding typings, users of the package cannot add untyped content just like with your approach. Even within the object, it is constrained to only accept shared types, which (in a way) is more strongly typed than in SharedModels.

**Comment by echarles** [comment]
> This PR will basically re-introduce modeldb under a different name (which unfortunately is also called "shared models"..).

I second @dmonad in the interpretation of this PR.

> This approach is fundamentally untyped and will make all the different packages again responsible for rendering the dom and also maintaining the structure of the shared Yjs document. These concerns should be separated. I highly recommend that JupyterLab maintains shared models in a single place, which handles all the complicated edge-cases of concurrency. Third-party packages should not get in contact with the CRDT itself because they could insert invalid data (or data that no other client can understand).

I agree with `Third-party packages should not get in contact with the CRDT itself because they could insert invalid data (or data that no other client can understand).` and I fear that this PR will make that mandatory at some point.

The shared-model looks now like providing basic structures where no notion of notebook, cell, metadata can be found, leaving the responsibility of that to the other JupyterLab 100+ (soon) packages.

Screenshot 2022-05-20 at 21 12 55

**Comment by echarles** [comment]
> After talking about this with several people in the team, I am clearly in favor of Carlos' approach in this PR.

@SylvainCorlay Which team are you referring to?

**Comment by SylvainCorlay** [comment]
> > This PR will basically re-introduce modeldb under a different name (which unfortunately is also called "shared models"..).
>
> I second @dmonad in the interpretation of this PR

Yes that is correct.

**Comment by afshin** [comment]
@echarles, I think Sylvain is referring to multiple conversations with me, @dmonad, @fcollonval, @davidbrochart , and @hbcarlos at various stages of this PR's lifecycle.

I think the crux of Carlos's approach is to resolve the duplication of notebook content in our original implementation. That problem needs to be resolved one way or another. 

It does sound like a dedicated conversation might be beneficial.

**Comment by echarles** [comment]
Has the team considered/discussed removing `modeldb` usage without touching the `shared-models`? This was the initial goal discussed some time ago, and more than one year after, it still sounds to me like the correct next step.

I still think there is a better way to move forward than what this PR does.

**Comment by SylvainCorlay** [comment]
> Has the team considered/discussed removing modeldb usage without touching the shared-models? This was the initial goal discussed some time ago, and more than one year after, it still sounds to me like the correct next step.
>
> I still think there is a better way to move forward than what this PR does.

Initial experiments in this direction proved deeply impractical. There are intrinsic reasons why this would break things such as the ones exposed above wrt metadata etc.

**Comment by SylvainCorlay** [comment]
Incidentally this *appears* to be fixing the data loss issue we had at the last lab team meeting.

**Comment by echarles** [comment]
Thx for the inputs. As discussed with @hbcarlos let's review this during the next Jupyter RTC community call next Monday, May 30, 17:00 – 18:00 CET.

**Comment by ellisonbg** [comment]
At the RTC meeting this morning, @hbcarlos walked us through the main points of this PR. Before diving into specific comments/questions I wanted to thank @hbcarlos for diving into this. This particular part of JupyterLab has always been technical complex and difficult. My sense is that this is very difficult to get right this layer right, that the details matter, and that the potential friction for JupyterLab and extension developer is significant. So @hbcarlos thank you for wading into the pile of ambiguity that is pushing us to better understand the problems and their possibly solutions.

**Comment by ellisonbg** [comment]
Continuing, @SylvainCorlay and I talked later on today to try and summarize the main points of discussion from the meeting. To help guide the discussion, we identified the core questions that we believe need to be answered:

1. In what package should models exist for notebooks, text files, cells, outputs, etc?
2. Should we use Yjs types diHrectly, or wrap them in a "shared types" API that is used as the public API instead of Yjs?
3. Should our models prevent schema-breaking document mutations?
4. How many layers of model APIs should we have?
5. How do we model highly nested documents, such as the nested metadata of notebooks?
6. How is model initialization handled?
7. Do we need model factories to allow the customization of models?
8. What the public model API that views and extensions will use?
9. How do we model non-shared state on documents?

We have a bunch of informal notes on these questions that probably need to be cleaned up before posting, but I wanted to at least post the questions, as those are useful for framing the discussion.

**Comment by davidbrochart** [comment]
> 1. In what package should models exist for notebooks, text files, cells, outputs, etc?

What about moving them to [jupyter_ydoc](https://github.com/jupyter-server/jupyter_ydoc)? I opened [an issue](https://github.com/jupyterlab/jupyterlab/issues/12603) some time ago. I makes sense to have all models, for any language, in the same repo/package (for keeping in sync and testing).

**Comment by echarles** [comment]
> What about moving them to [jupyter_ydoc](https://github.com/jupyter-server/jupyter_ydoc)? I opened https://github.com/jupyterlab/jupyterlab/issues/12603 some time ago. I makes sense to have all models, for any language, in the same repo/package (for keeping in sync and testing).

That is an intriguing and brilliant approach. Love the initiative and curious to see what others think about that.

> Should we use Yjs types diHrectly, or wrap them in a "shared types" API that is used as the public API instead of Yjs?

I am strongly in fa or to not use the Y.js types directly but exposing them in whatever shared types. So no public API to the Y.js types.

> How many layers of model APIs should we have?

I guess we have to deal with the existing implementations. Instead of forcing a specific number, I would favor a criteria which would be "how can we smoothly evolve step by step without too many fundamental changes and without impacting too much the extension authors".

> How is model initialization handled?

As I have quickly said yesterday, that has to be looked across client+server eyes, hence the @davidbrochart proposal makes sense.

**Comment by SylvainCorlay** [comment]
In the previous comment, I addressed the parallel hierarchy of classes that was introduced in the first iteration on RTC, and which is removed in this PR.

Another point that this PR touches is the **modelling of metadata**.

In the earlier versions of Lab, metadata was modelled using nested observable types from ModelDB. This allowed extensions to observe individual metadata keys. In the shared-models approach of Kevin, the metadata can only be set and fetched globally, but not observed atomically. (This choice was made - in my understanding - to address the situation in which two different attributes in the metadata must be synced, in which case two users modifying different keys could end up in the auto-merge giving an inconsistent result.) In this PR, Carlos allows for metadata to be observed more granularly to continue supporting the use case of extensions that rely on observing the metadata.

My opinion is that we need this granular modelling of metadata to preserve the use cases of existing extensions. Note that this choice could also be made on top of the approach proposed by Kevin.

The question of the granularity of the modelling of metadata is the subject of the team-compass issue https://github.com/jupyterlab/team-compass/issues/150.

**Comment by JohanMabille** [comment]
> Remove ModelDB and the shared state from NotebookModel, CodeCellModel, etc. and pair them each with a shared model: YNotebook, YCodeCell, etc.

I like that this PR removes the parallel hierarchies of classes introduced in the first iteration on RTC, which are usually a code smell. Besides, the simplicity of keeping all the shared types in a single package is kind of an illusion to me: each time we want to add a new model, we need to add a new class in the common package too, duplicating (and having to maintain) the relations between the different classes of the hierarchies.

Another point is now that we need a CRDT implementation in the backend, we can definitely not say that a particular implementation can be used as a "schema". I
understand that fully consistent schemes for CRDTs is a problem that has not been solved yet. However, we should be able to test different implementations
against a kind of degraded schema, just like we can test the implementation of a protocol.

**Comment by SylvainCorlay** [comment]
Regarding the argument of line count, here the rough breakdown of the diff (+2900 lines of code)

1) ~ 1000 lines of code for the tests
2) A few hundreds of lines of code for the docstrings
3) ~ 2900 lines of code for the Yjs abstraction layer (which is required to not expose Yjs types while having a granular modelling of the metadata).
4) Lines removed for the removal of shared models and old ModelDB.

I think that the overall diff is justified, and essentially accounts for the Yjs abstraction layer.

File: packages/codeeditor/src/editor.ts [hunk 1]
```diff
@@ -227,123 +234,81 @@ export namespace CodeEditor {
...
-        // update local modeldb
-        // @todo also change metadata
-        this.value.text = sharedModel.getSource();
-      }
-      this.sharedModel.changed.disconnect(this._onSharedModelChanged, this);
-      // clone model retrieve a shared (not standalone) model
-      this.sharedModel = sharedModel;
-      this.sharedModel.changed.connect(this._onSharedModelChanged, this);
-      this._sharedModelSwitched.emit(true);
+    get type(): nbformat.CellType {
```
Comment (SylvainCorlay) [inline packages/codeeditor/src/editor.ts:262]: From @afshin: We probably should not depend on `nbformat` in this place. Do we need the `type` method here.

File: packages/shared-models/package.json [hunk 2]
```diff
@@ -37,7 +37,7 @@
     "watch": "tsc -b --watch"
   },
   "dependencies": {
-    "@jupyterlab/nbformat": "^4.0.0-alpha.11",
+    "@lumino/algorithm": "^1.9.1",
```
Comment (SylvainCorlay) [inline packages/shared-models/package.json:40]: Maybe we should call that package `shared-types`.

Comment (afshin) [reply packages/shared-models/package.json:40]: I think this proposed `@jupyterlab/shared-types` package should export a single namespace called `SharedTypes` with both interfaces and concrete implementations, e.g., `SharedTypes.List` and `SharedTypes.IList`. This will minimize the number of individually named items a consumer of the library will have to `import` to use it.

**Comment by afshin** [comment]
@hbcarlos, @dmonad, @sylvaincorlay, and I were at Monday's RTC bi-weekly call and we discussed closing this PR and moving forward with https://github.com/jupyterlab/jupyterlab/pull/12695 instead of this PR.

Here are some aspects of Monday's conversation:

1. From a naive point of view, when I imagine how Yjs CRDT primitives could be used to build RTC in a Jupyter front-end built from the ground up, I think my instincts would be to use Yjs "natively" in JupyterLab but:
  - This is super disruptive and would require major refactors everywhere.
  - Kevin made a compelling argument that doing this would yield bad outcomes anywhere you have, say, two attributes that need to be in sync and tightly coupled. His example was something like an `{x, y}` coordinate pair where unless you keep `x` and `y` in sync you end up with meaningless values. So he suggests using Yjs primitives in the "core" code but allowing your clients to interact with APIs you author
2. In lieu of the approach above, we could consider writing an abstraction layer to allow us to one day switch to another CRDT. But the CodeMirror `IEditor` abstraction has proved more a barrier than an asset. Carlos's shared types work
plus my experimental refactor (both contained in this PR) persuaded me that this is just not doable because without refactoring other core parts of JupyterLab, it will remain a leaky abstraction.

During Monday's call,  Kevin agreed that moving the `Y` models into their semantic namespace packages `YNotebook` in `@jupyterlab/notebook`, for example, and re-exporting them in a `@jupyterlab/shared-models` meta-package is a
reasonable change to apply to his approach.  (cf. https://github.com/jupyterlab/team-compass/issues/149)

We also talked about making sure the `metadata` implementation allows for key-level access. (cf. https://github.com/jupyterlab/team-compass/issues/150)

We asked Carlos to see which of his tests might be applicable to Kevin's PR. And finally, instantiation catastrophic data loss has been fixed in this PR, so Carlos and Kevin discussed the fix and made sure it is fixed in Kevin's PR.

Thanks for all the hard work, experimentation, and iterating on this. It has clearly been a lot of hard work and appreciated effort.

**Comment by echarles** [comment]
> During Monday's call, Kevin agreed that moving the Y models into their semantic namespace packages YNotebook in @jupyterlab/notebook, for example, and re-exporting them in a @jupyterlab/shared-models meta-package is a
reasonable change to apply to his approach. (cf. https://github.com/jupyterlab/team-compass/issues/149)

Do I read this as "shared-models will have a dependency on notebook, cells... packages"? If such it is not have real value.
